### ISSUE TITLE
Mux H264 video frames into fragmented MP4 

### ISSUE THREAD

**Issue by david7482** [gpac/mp4box.js#68]
Hi all,

If I got H264 AnnexB frames streaming from WebSocket or WebRTC data channel, could I use mp4box to mux them into fragmented MP4 and make browser to play with MSE ?
Because I only got examples from non-fragmented MP4 to fragmented MP4, is it doable ?

Thanks

**Comment by cconcolato** [comment]
This is currently not possible. It's not yet on the TODO list. In the meantime, you may have a look at https://github.com/dailymotion/hls.js which creates MP4 fragments from MPEG-2 TS.

**Comment by david7482** [comment]
Thanks for your comment :)

My original thought was if MP4Box is able to remux non-fragment MP4 to a fragment one, there should be somewhere it got the raw samples then do the remux job. So, I could try to expose this part to the interface, then it might work.

In your point of view, is it doable ?
Thanks !

**Comment by cconcolato** [comment]
In non-fragmented MP4, just like in fragmented MP4, H.264 frames are not stored in Annex B format. Currently, MP4Box.js does not handle Annex B frames. It uses length fields and no start codes. The [HEVC Image Viewer sample application](https://github.com/gpac/mp4box.js/tree/master/test/hevcimageviewer) contains some code to do the conversion, but for HEVC. So this is doable with H264 as well. Then, MP4Box.js does not deal specifically with SPS/PPS, so if you would want to create 'avc1' streams (not 'avc3') you would have to provide them to MP4Box.js. 

In summary, yes in theory this would be doable. In practice, this would require some work.

**Comment by cousine** [comment]
Hi,

Thanks for the detailed info in this thread !

Right now I'm facing a similar situation, I have raw h264 and aac data that are being transported over webrtc and I'm researching if its possible to use mp4boxjs to create a segmented container for MSE on the fly, as far as I know the h264 is in avcc not annex, I have the same line of thought as david, I'd love to expose part of the api to feed mp4box the ray data.

I'd love to contribute this feature if its doable, so some pointers on where to start would be great :smile:

**Comment by cconcolato** [comment]
@cousine In your case, since the data is directly compatible with the MP4 storage. It's easier. We could easily add something like:
var track = mp4box.addTrack(mime, options)
and 
track.addSample(data, options)
The only question is whether you want to produce avc1 or avc3. In avc1, the addTrack option would need to have the PPS/SPS otherwise the addSample would have them. What do you think?

**Comment by david74chou** [comment]
Hi @cconcolato 

It's good to know we could have these APIs.

I think avc3 would be better since we could have SPS/PPS in each fragment. 
Then, we could achieve live streaming with adaptive resolution through MSE.

Thanks

**Comment by david74chou** [comment]
Hi @cconcolato

It's good to know we could have these APIs.

I think avc3 would be better since we could have SPS/PPS in each fragment. 
Then, we could achieve live streaming with adaptive resolution through MSE.

Thanks

**Comment by david7482** [comment]
Hi @cconcolato

It's good to know we could have these APIs.

I think avc3 would be better since we could have SPS/PPS in each fragment. 
Then, we could achieve live streaming with adaptive resolution through MSE.

Thanks

**Comment by cousine** [comment]
Yea avc3 would be better indeed, I'll have a look at it sometime this week and probably going to dedicate some time starting next week, will keep this issue updated and will create a PR once ready.

Thanks for the help guys :smile:

**Comment by 131** [comment]
@cousine , @david7482 , the exact same research (playing live h264) drove me in the opposite direction. I ended up using a raw h264 javascript decoded (via emscripten) and drop MSE/ mp4 live boxing . Take a look at https://github.com/131/h264-live-player

Yet (10 months later, and regardless of my project success) i'm still searching for a mp4box.js/MSE alternative.

**Comment by rbouqueau** [comment]
> I think avc3 would be better since we could have SPS/PPS in each fragment. 

Supporting avc3 in mp4box.js indeed looks easier than adding AnnexB parsing.
@cousine Could be a great PR, looking forward to it :)

**Comment by mbivol10** [comment]
@131 why not try and code this from scratch. Start a repo and lets work on it together I've also been looking for muxing h264 into mp4 on the fly. Unfortunately everything leads to either this or ffmpeg and both seem out of option since they are unrealistic for outside testing purposes. So one of these days start on writing an on the fly container maker.

**Comment by parsifal-47** [comment]
Now there is a tool for that and I even made it work for me in Chrome, FF and Safari
https://github.com/videojs/mux.js

**Comment by rbouqueau** [comment]
It indeed seems this project has an Annex B H264 parser.

**Comment by ChrisChiasson** [comment]
@parsifal-47 @rbouqueau can you explain a little more about how mux.js helps? (as it relates to either @david7482 or @cousine 's situations they've explained) Just trying to understand. Respectfully, Chris

**Comment by cconcolato** [comment]
See #230 or #185
